libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp

```diff
@@ -0,0 +1,228 @@
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Common/AP_Common.h>
+#include <AP_Math/AP_Math.h>
+#include <AP_Notify/AP_Notify.h>
+#include "AP_BattMonitor.h"
+#include "AP_BattMonitor_SMBus_SUI.h"
+
+extern const AP_HAL::HAL& hal;
+
+// button held down for 3 intervals will cause a power off event
+#define BATTMONITOR_SMBUS_SUI_BUTTON_DEBOUNCE 3
+
+#define REG_CELL_VOLTAGE        0x28
+#define REG_CURRENT             0x2a
+
+// maximum number of cells that we can read data for
+#define SUI_MAX_CELL_READ       4
+
+// Constructor
+AP_BattMonitor_SMBus_SUI::AP_BattMonitor_SMBus_SUI(AP_BattMonitor &mon,
+        AP_BattMonitor::BattMonitor_State &mon_state,
+        AP_BattMonitor_Params &params,
+        AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev,
+        uint8_t _cell_count)
+    : AP_BattMonitor_SMBus(mon, mon_state, params, std::move(dev)),
+      cell_count(_cell_count)
+{
+    _pec_supported = false;
+    _dev->set_retries(2);
+}
+
+void AP_BattMonitor_SMBus_SUI::init(void)
+{
+    AP_BattMonitor_SMBus::init();
+    if (_dev && timer_handle) {
+        // run twice as fast for two phases
+        _dev->adjust_periodic_callback(timer_handle, 50000);
+    }
+}
+
+/*
+  average over an array of the number of readable cells
+ */
+float AP_BattMonitor_SMBus_SUI::avg(const float *p) const
+{
+    float sum = 0;
+    const uint8_t n = MIN(SUI_MAX_CELL_READ, cell_count);
+    for (uint8_t i=0; i<n; i++) {
+        sum += p[i];
+    }
+    return sum / n;
+}
+
+void AP_BattMonitor_SMBus_SUI::timer()
+{
+    uint32_t tnow = AP_HAL::micros();
+
+    // we read in two phases as the device can stall if you read
+    // current too rapidly after voltages
+    phase_voltages = !phase_voltages;
+
+    if (phase_voltages) {
+        read_cell_voltages();
+        read_remaining_capacity();
+        return;
+    }
+
+    // read current
+    int32_t current_ma;
+    if (read_block_bare(REG_CURRENT, (uint8_t *)&current_ma, sizeof(current_ma))) {
+        _state.current_amps = -current_ma * 0.001;
+        _state.last_time_micros = tnow;
+    }
+
+    read_full_charge_capacity();
+
+    read_remaining_capacity();
+
+    // timeout after 5 seconds
+    if ((tnow - _state.last_time_micros) > AP_BATTMONITOR_SMBUS_TIMEOUT_MICROS) {
```

Comment:
This is pointless here, you never return before you get here, you need to have a check that you actually managed to do . the capacity calculations before this, otherwise the driver might as well always be healthy and just not manipulate this.

Reply:
fixed